single_cell_RNAseq_snake/bin/de_pool.R

```diff
@@ -0,0 +1,90 @@
+# This script depools demultiplexed single cell RNA seq data
+
+load_seurat_objects <- function(file_paths) {
+    seurat_objects <- list()
+
+    for (file_path in file_paths) {
+        if (file.exists(file_path)) {
+            seurat_objects[[basename(file_path)]] <- readRDS(file_path)
+        } else {
+            stop("File not found: ", file_path)
+        }
+    }
+    return(seurat_objects)
+}
+
+
+save_seurat_objects <- function(output_dir, seurat_list) {
+  # Check if the provided list is actually a list of Seurat objects
+  if (!is.list(seurat_list)) {
+    stop("The provided 'seurat_list' is not a list.")
+  }
+
+  # Ensure output directory exists
+  if (!dir.exists(output_dir)) {
+      dir.create(output_dir, recursive = TRUE)
+  }
+
+  # Iterate through each Seurat object in the list
+  for (id in names(seurat_list)) {
+    obj <- seurat_list[[id]]
+
+    # Define the filename using the output directory and the ID
+    filename <- file.path(output_dir, paste0(id, ".rds"))
+
+    # Save the Seurat object
+    saveRDS(obj, file = filename)
+  }
+}
+
+depool_by_id <- function(seurat_objects) {
+  depooled_objs <- list()
+  
+  for (seurat_obj in seurat_objects) {
+    # Extract metadata
+    metadata <- seurat_obj@meta.data
+    
+    # Split BEST.GUESS values and extract first for SNG, then store in 'sample_id'
+    metadata$sample_id <- sapply(strsplit(as.character(metadata$BEST.GUESS), ","), function(x) x[1])
+
+    # Update the metadata in the Seurat object
+    seurat_obj <- Seurat::AddMetaData(object = seurat_obj, metadata = metadata)
```

Comment:
Some potential simplification here

Curious of Emily's thoughts here since you suggested the AddMetaData approach, but the way I would normally do this is the one line of:

```
seurat_obj$sample_id <- sapply(strsplit(as.character(seurat_obj$BEST.GUESS), ","), function(x) x[1])
```

For both assignment and retrieval, `seurat_obj$<thing>` sets or grabs a metadata column named `<thing>`.

Even down below in line 58, you can do `colnames(seurat_obj)[<the logical you have but using 'seurat_obj' instead of 'metadata'>]`, so from my perspective there's no reason to be pulling out the `metadata <- seurat_obj@meta.data` specifically.

Reply:
I think Dan's way is simpler and correct! For line 51 it looks like you're adding the entire metadata not just that column.

I just avoid reassigning the entire metadata slot `sobj@meta.data <- metadata` unless you are being careful not to lose rownames.